File: packages/csharp/Kitsy.Cnos.Gcp/GcpSecretManagerProvider.cs
using System;
using System.Collections.Generic;
using System.Text;
using Google.Api.Gax.ResourceNames;
using Google.Cloud.SecretManager.V1;
using Grpc.Core;
using Kitsy.Cnos;

namespace Kitsy.Cnos.Gcp
{
    /// <summary>
    /// CNOS vault provider for Google Cloud Secret Manager.
    /// Register via <c>new SecretVaultProviderFactory("gcp", (id, def) => new GcpSecretManagerProvider(projectId))</c>.
    /// </summary>
    public sealed class GcpSecretManagerProvider : ISecretVaultProvider
    {
        private readonly string _projectId;
        private SecretManagerServiceClient? _client;

        /// <param name="projectId">GCP project ID.</param>
        public GcpSecretManagerProvider(string projectId)
        {
            _projectId = projectId ?? throw new ArgumentNullException(nameof(projectId));
        }

        public void Authenticate(VaultAuthConfig auth)
        {
            _client = SecretManagerServiceClient.Create();
        }

        public Dictionary<string, object?> BatchGet(IReadOnlyList<string> refs)
        {
            var result = new Dictionary<string, object?>(StringComparer.Ordinal);
            foreach (string @ref in refs)
            {
                object? value = GetSingle(@ref);
                if (value != null) result[@ref] = value;
            }
            return result;
        }

        public object? Get(string reference) => GetSingle(reference);

        private object? GetSingle(string secretId)
        {
            if (_client == null)
                throw new CnosError("cnos: GcpSecretManagerProvider.Authenticate must be called before reading secrets");
            try
            {
                string resourceName = $"projects/{_projectId}/secrets/{secretId}/versions/latest";
                AccessSecretVersionResponse response = _client.AccessSecretVersion(resourceName);
                return response.Payload.Data.ToStringUtf8();
            }
            catch (RpcException ex) when (ex.StatusCode == StatusCode.NotFound)
            {
                return null;
            }
            catch (Exception ex) when (ex is not CnosError)
            {
                throw new CnosError($"cnos: GCP Secret Manager error for \"{secretId}\": {ex.Message}", ex);
            }
        }

        /// <summary>Returns a <see cref="SecretVaultProviderFactory"/> for this provider.</summary>
        public static SecretVaultProviderFactory Factory(string projectId) =>
            new SecretVaultProviderFactory("gcp", (_, _) => new GcpSecretManagerProvider(projectId));
    }
}

File: packages/php/cnos-gcp/src/GcpSecretManagerProvider.php
<?php

declare(strict_types=1);

namespace Kitsy\Cnos\Gcp;

use Google\Cloud\SecretManager\V1\SecretManagerServiceClient;
use Kitsy\Cnos\CnosError;
use Kitsy\Cnos\SecretVaultProvider;
use Kitsy\Cnos\SecretVaultProviderFactory;
use Kitsy\Cnos\VaultAuthConfig;
use Kitsy\Cnos\VaultDefinition;

class GcpSecretManagerProvider implements SecretVaultProvider
{
    private ?SecretManagerServiceClient $client;
    private readonly string $projectId;
    private readonly string $version;

    public function __construct(
        private readonly string $vaultId,
        VaultDefinition $definition,
        ?SecretManagerServiceClient $client = null,
    ) {
        $cfg            = $definition->auth->config;
        $this->projectId = (string) ($cfg['project'] ?? $cfg['projectId'] ?? getenv('GOOGLE_CLOUD_PROJECT') ?: '');
        $this->version   = (string) ($cfg['version'] ?? 'latest');
        $this->client    = $client;
    }

    public function authenticate(VaultAuthConfig $auth): void
    {
        if ($this->client !== null) return;

        if (!class_exists(SecretManagerServiceClient::class)) {
            throw new CnosError(
                'cnos-gcp: google/cloud-secret-manager is required. '
                . 'Run: composer require google/cloud-secret-manager'
            );
        }
        if ($this->projectId === '') {
            throw new CnosError(
                "cnos-gcp: project ID is required. Set GOOGLE_CLOUD_PROJECT or vault auth config.project"
            );
        }
        $this->client = new SecretManagerServiceClient();
    }

    public function batchGet(array $refs): array
    {
        $result = [];
        foreach ($refs as $ref) {
            $val = $this->getOne((string) $ref);
            if ($val !== null) {
                $result[$ref] = $val;
            }
        }
        return $result;
    }

    public function get(string $ref): mixed
    {
        return $this->getOne($ref);
    }

    private function getOne(string $ref): ?string
    {
        // ref format: "secret-name" or "projects/…/secrets/name/versions/latest"
        if (!str_starts_with($ref, 'projects/')) {
            $ref = "projects/{$this->projectId}/secrets/{$ref}/versions/{$this->version}";
        }
        try {
            $response = $this->client->accessSecretVersion($ref);
            return $response->getPayload()?->getData() ?? null;
        } catch (\Google\ApiCore\ApiException $e) {
            if ($e->getStatus() === 'NOT_FOUND') return null;
            throw new CnosError("cnos-gcp: Secret Manager error for \"{$ref}\": {$e->getMessage()}", $e);
        }
    }

    public static function factory(?SecretManagerServiceClient $client = null): SecretVaultProviderFactory
    {
        return new SecretVaultProviderFactory(
            provider: 'gcp-secret-manager',
            create: static fn(string $id, VaultDefinition $def): self => new self($id, $def, $client),
        );
    }
}
